src/app/core/common-components/json-editor/json-editor.component.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  Component,
+  ElementRef,
+  EventEmitter,
+  Output,
+  ViewChild,
+  Optional,
+  Self,
+} from "@angular/core";
+import { CommonModule } from "@angular/common";
+import { Content, createJSONEditor } from "vanilla-jsoneditor/standalone.js";
+import { AlertService } from "../../alerts/alert.service";
+import { CustomFormControlDirective } from "../../common-components/basic-autocomplete/custom-form-control.directive";
+import { NgControl, NgForm, FormGroupDirective } from "@angular/forms";
+import { ErrorStateMatcher } from "@angular/material/core";
+import { MatFormFieldControl } from "@angular/material/form-field";
+
+/**
+ * Component for editing JSON data.
+ */
+@Component({
+  selector: "app-json-editor",
+  standalone: true,
+  imports: [CommonModule],
+  templateUrl: "./json-editor.component.html",
+  styleUrl: "./json-editor.component.scss",
+  providers: [
+    { provide: MatFormFieldControl, useExisting: JsonEditorComponent },
+  ],
+})
+export class JsonEditorComponent extends CustomFormControlDirective<object> {
+  @Output() isValidChange = new EventEmitter<boolean>();
```

Comment:
Ideally we would mark the FormControl as "invalid", rather than introducing our own isValidChange output ... But I am not sure myself how to implement that in a nice way within this JsonEditorComponent.

If you, @Ayush8923 , maybe with any of your seniors, have an idea for it. Let's try that. Otherwise, we'll stick with the current version, which works alright.

Reply:
@sleidig I have updated the code to handle everything using the formControl. Additionally, if the JSON is invalid, the corresponding `formControl` will be marked as invalid, and the `Save` button will be disabled. Please review the changes and let me know if any further adjustments are required.

You can test the functionality in this PR deployment link: https://pr-2785.aam-digital.net/

Also, if in this Notification JSON editor, [PR](https://github.com/Aam-Digital/ndb-core/pull/2785) changes look good then can we merge these changes on the final notification branch because I made some cleanups in this PR?

Reply:
Okay, I have investigated the cause a bit further still:

Previously, you had defined `@Input() formControl` in JsonEditorComponent. This made it just a simple Input, unrelated to the ReactiveForms module of Angular and the special CustomFormControl's injected NgControl .

Now, we have removed the Input . `<app-json-editor [formControl]="jsonDataControl"` would now actually complain that there is no "formControl" input. But as we import the ReactiveFormsModule here (in the JsonEditorDialog), then this "formControl" directive becomes available from the ReactiveFormsModule and links into angular's system ... which in turn makes it available as NgControl through the dependency injection